File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/label/LabelTest.java
package openfoodfacts.github.scrachx.openfood.models.entities.label;

import org.greenrobot.greendao.DaoException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import openfoodfacts.github.scrachx.openfood.models.DaoSession;

import static com.google.common.truth.Truth.assertThat;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_FRENCH;
import static openfoodfacts.github.scrachx.openfood.models.entities.label.LabelNameTestData.LABEL_NAME_EN;
import static openfoodfacts.github.scrachx.openfood.models.entities.label.LabelNameTestData.LABEL_NAME_FR;
import static openfoodfacts.github.scrachx.openfood.models.entities.label.LabelNameTestData.LABEL_TAG;
import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Tests for {@link Label}
 */
@RunWith(MockitoJUnitRunner.class)
public class LabelTest {
    @Mock
    private DaoSession mockDaoSession;
    @Mock
    private LabelDao mockLabelDao;
    @Mock
    private LabelNameDao mockLabelNameDao;
    private Label mLabel;

    @Before
    public void setup() {
        when(mockDaoSession.getLabelDao()).thenReturn(mockLabelDao);
        when(mockDaoSession.getLabelNameDao()).thenReturn(mockLabelNameDao);
        LabelName labelName1 = new LabelName(LABEL_TAG, LANGUAGE_CODE_ENGLISH, LABEL_NAME_EN);
        LabelName labelName2 = new LabelName(LABEL_TAG, LANGUAGE_CODE_FRENCH, LABEL_NAME_FR);
        when(mockLabelNameDao._queryLabel_Names(any()))
            .thenReturn(Arrays.asList(labelName1, labelName2));
        mLabel = new Label();
    }

    @Test
    public void getNamesWithNullNamesAndNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mLabel.getNames());
    }

    @Test
    public void getNamesWithNullNamesAndNonNullDaoSession_getsNamesFromLabelNameDao() {
        mLabel.__setDaoSession(mockDaoSession);
        List<LabelName> labelNames = mLabel.getNames();

        assertThat(labelNames).hasSize(2);

        LabelName labelName1 = labelNames.get(0);
        assertThat(labelName1.getLabelTag()).isEqualTo(LABEL_TAG);
        assertThat(labelName1.getLanguageCode()).isEqualTo(LANGUAGE_CODE_ENGLISH);
        assertThat(labelName1.getName()).isEqualTo(LABEL_NAME_EN);

        LabelName labelName2 = labelNames.get(1);
        assertThat(labelName2.getLabelTag()).isEqualTo(LABEL_TAG);
        assertThat(labelName2.getLanguageCode()).isEqualTo(LANGUAGE_CODE_FRENCH);
        assertThat(labelName2.getName()).isEqualTo(LABEL_NAME_FR);
    }

    @Test
    public void deleteWithNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mLabel.delete());
    }

    @Test
    public void deleteWithNonNullDaoSession_callsDeleteOnLabelDao() {
        mLabel.__setDaoSession(mockDaoSession);
        mLabel.delete();
        verify(mockLabelDao).delete(mLabel);
    }

    @Test
    public void refreshWithNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mLabel.refresh());
    }

    @Test
    public void refreshWithNonNullDaoSession_callsRefreshOnLabelDao() {
        mLabel.__setDaoSession(mockDaoSession);
        mLabel.refresh();
        verify(mockLabelDao).refresh(mLabel);
    }

    @Test
    public void updateWithNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mLabel.update());
    }

    @Test
    public void updateWithNonNullDaoSession_callsUpdateOnLabelDao() {
        mLabel.__setDaoSession(mockDaoSession);
        mLabel.update();
        verify(mockLabelDao).update(mLabel);
    }

    @Test
    public void resetNames_callsGetLabelNameDao() {
        mLabel.__setDaoSession(mockDaoSession);
        mLabel.resetNames();
        mLabel.getNames();
        verify(mockDaoSession).getLabelNameDao();
    }
}

File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/label/LabelTest.kt
package openfoodfacts.github.scrachx.openfood.models.entities.label

import com.google.common.truth.Truth
import openfoodfacts.github.scrachx.openfood.models.DaoSession
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_FRENCH
import openfoodfacts.github.scrachx.openfood.models.entities.label.LabelNameTestData.LABEL_NAME_EN
import openfoodfacts.github.scrachx.openfood.models.entities.label.LabelNameTestData.LABEL_NAME_FR
import openfoodfacts.github.scrachx.openfood.models.entities.label.LabelNameTestData.LABEL_TAG
import org.greenrobot.greendao.DaoException
import org.junit.Assert
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.whenever
import org.mockito.quality.Strictness

/**
 * Tests for [Label]
 */
@ExtendWith(MockitoExtension::class)
@MockitoSettings(strictness = Strictness.LENIENT)
class LabelTest {
    @Mock
    private lateinit var mockDaoSession: DaoSession

    @Mock
    private lateinit var mockLabelDao: LabelDao

    @Mock
    private val mockLabelNameDao: LabelNameDao? = null

    private lateinit var mLabel: Label

    @BeforeEach
    fun setup() {
        whenever(mockDaoSession.labelDao) doReturn mockLabelDao
        whenever(mockDaoSession.labelNameDao) doReturn mockLabelNameDao
        val labelName1 = LabelName(LABEL_TAG, LANGUAGE_CODE_ENGLISH, LABEL_NAME_EN)
        val labelName2 = LabelName(LABEL_TAG, LANGUAGE_CODE_FRENCH, LABEL_NAME_FR)
        whenever(mockLabelNameDao!!._queryLabel_Names(ArgumentMatchers.any())) doReturn listOf(labelName1, labelName2)
        mLabel = Label()
    }

    @Test
    fun getNamesWithNullNamesAndNullDaoSession_throwsDaoException() {
        Assert.assertThrows(DaoException::class.java) { mLabel.names }
    }

    @Test
    fun getNamesWithNullNamesAndNonNullDaoSession_getsNamesFromLabelNameDao() {
        mLabel.__setDaoSession(mockDaoSession)
        val labelNames = mLabel.names
        Truth.assertThat(labelNames).hasSize(2)
        val labelName1 = labelNames[0]
        Truth.assertThat(labelName1!!.labelTag).isEqualTo(LABEL_TAG)
        Truth.assertThat(labelName1.languageCode).isEqualTo(LANGUAGE_CODE_ENGLISH)
        Truth.assertThat(labelName1.name).isEqualTo(LABEL_NAME_EN)
        val labelName2 = labelNames[1]
        Truth.assertThat(labelName2!!.labelTag).isEqualTo(LABEL_TAG)
        Truth.assertThat(labelName2.languageCode).isEqualTo(LANGUAGE_CODE_FRENCH)
        Truth.assertThat(labelName2.name).isEqualTo(LABEL_NAME_FR)
    }

    @Test
    fun deleteWithNullDaoSession_throwsDaoException() {
        Assert.assertThrows(DaoException::class.java) { mLabel.delete() }
    }

    @Test
    fun deleteWithNonNullDaoSession_callsDeleteOnLabelDao() {
        mLabel.__setDaoSession(mockDaoSession)
        mLabel.delete()
        verify(mockLabelDao)!!.delete(mLabel)
    }

    @Test
    fun refreshWithNullDaoSession_throwsDaoException() {
        Assert.assertThrows(DaoException::class.java) { mLabel.refresh() }
    }

    @Test
    fun refreshWithNonNullDaoSession_callsRefreshOnLabelDao() {
        mLabel.__setDaoSession(mockDaoSession)
        mLabel.refresh()
        verify(mockLabelDao)!!.refresh(mLabel)
    }

    @Test
    fun updateWithNullDaoSession_throwsDaoException() {
        Assert.assertThrows(DaoException::class.java) { mLabel.update() }
    }

    @Test
    fun updateWithNonNullDaoSession_callsUpdateOnLabelDao() {
        mLabel.__setDaoSession(mockDaoSession)
        mLabel.update()
        verify(mockLabelDao)!!.update(mLabel)
    }

    @Test
    fun resetNames_callsGetLabelNameDao() {
        mLabel.__setDaoSession(mockDaoSession)
        mLabel.resetNames()
        mLabel.names
        verify(mockDaoSession)!!.labelNameDao
    }
}